contrib/mypy/src/python/pants/contrib/mypy/tasks/mypy_task.py

```diff
@@ -0,0 +1,118 @@
+from pants.backend.python.tasks.python_task import PythonTask
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnit, WorkUnitLabel
+from pants.util.contextutil import temporary_file_path
+from pants.util.process_handler import subprocess
+from pex.pex_info import PexInfo
+
+
+class MypyTask(PythonTask):
+  """Invoke the mypy static type analyzer for Python."""
+
+  _PYTHON_SOURCE_EXTENSION = '.py'
+
+  @classmethod
+  def register_options(cls, register):
+    register('--mypy-version', default='0.550', help='The version of mypy to use')
+
+  @classmethod
+  def supports_passthru_args(cls):
+    return True
+
+  def find_py3_interpreter(self):
+    interpreters = self._interpreter_cache.setup(filters=['>=3'])
+    return min(interpreters) if interpreters else None
+
+  @staticmethod
+  def is_non_synthetic_python_target(target):
+    return not target.is_synthetic and isinstance(target, (PythonLibrary, PythonBinary, PythonTests))
+
+  @staticmethod
+  def is_python_target(target):
+    return isinstance(target, (PythonTarget,))
+
+  def _calculate_python_sources(self, targets):
+    """Generate a set of source files from the given targets."""
+    python_eval_targets = filter(self.is_non_synthetic_python_target, targets)
+    sources = set()
+    for target in python_eval_targets:
+      sources.update(
+        source for source in target.sources_relative_to_buildroot()
+        if os.path.splitext(source)[1] == self._PYTHON_SOURCE_EXTENSION
+      )
+    return list(sources)
+
+  def _collect_source_roots(self, targets):
+    # Collect the set of directories in which there are Python sources (whether part of
+    # the target roots or transitive dependencies.)
+    source_roots = set()
+    for target in self.context.targets(self.is_python_target):
+      if not target.has_sources(self._PYTHON_SOURCE_EXTENSION):
+        continue
+      source_roots.add(target.target_base)
+    return source_roots
+
+  def _run_mypy(self, py3_interpreter, mypy_args, **kwargs):
+    pex_info = PexInfo.default()
+    pex_info.entry_point = 'mypy'
+    chroot = self.cached_chroot(interpreter=py3_interpreter, pex_info=pex_info, targets=[],
+      extra_requirements=[PythonRequirement('mypy=={}'.format(self.get_options().mypy_version))])
+    pex = chroot.pex()
+    return pex.run(mypy_args, **kwargs)
+
+  def execute(self):
+    py3_interpreter = self.find_py3_interpreter()
+    if not py3_interpreter:
+      raise TaskError('Unable to find a Python 3.x interpreter (required for mypy).')
+
+    sources = self._calculate_python_sources(self.context.target_roots)
+    if not sources:
+      self.context.log.debug('No Python sources to check.')
+      return
+
+    # Determine interpreter used by the sources so we can tell mypy.
+    interpreter_for_targets = self._interpreter_cache.select_interpreter_for_targets(self.context.target_roots)
```

Comment:
@tdyas : I'm deprecating PythonTask, as it's part of the old pipeline (should have noticed you were relying on it in this change, my bad). So quick question - why are you selecting the interpreter only for self.context.target_roots rather than all targets (later you collect source roots for all targets, which I assume means that mypy operates on all of them)? Does this matter, or can I have this select an interpreter for all targets (and then use the already-computed product for that)?

Reply:
I chose to use `target_roots` since I only wanted `mypy` to operate on the targets passed on the `pants` command-line. I'm not wedded to that idea though. What is expected of a Pants code linting task?

As for switching to another Python pipeline, just point me at what I should be using.

Reply:
I'm still figuring out how this would fit into the new pipeline, since it has some unusual usage patterns (e.g., having to select two interpreters, one for running mypy and one for discovering the python version of the input files).